apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,520 @@
+    }
+
+    async fn inner_start<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        spawn: bool,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+        listener: Option<TcpListener>,
+    ) -> Self {
+        if update {
+            info!("Running in update mode ⬆️");
+        } else if offline {
+            info!("Running in offline mode ⛔️");
+        }
+
+        let snapshot_file = snapshot_path.as_ref();
+        let snapshots: Vec<Snapshot> = load(snapshot_file).unwrap_or(vec![]);
+
+        let http_service = HttpClientService::new(
+            "test",
+            rustls::ClientConfig::builder()
+                .with_safe_defaults()
+                .with_native_roots()
+                .with_no_client_auth(),
+            Client::builder().build(),
+        )
+        .expect("can create a HttpService");
+        let app = Router::new()
+            .route("/", any(root_handler))
+            .route("/*path", any(handler))
+            .with_state(SnapshotServerState {
+                client: http_service,
+                base_url: base_url.clone(),
+                snapshots: Arc::new(Mutex::new(snapshots.clone())),
+                snapshot_file: Box::from(snapshot_file),
+                offline,
+                update,
+                include_headers,
+            });
+        let listener = listener.unwrap_or(
+            TcpListener::bind("127.0.0.1:0")
+                .expect("Failed to bind an OS port for snapshot server"),
+        );
+        let local_address = listener
+            .local_addr()
+            .expect("Failed to get snapshot server address.");
+        info!(
+            "Snapshot server listening on port {:?}",
+            local_address.port()
+        );
+        if spawn {
+            tokio::spawn(async move {
+                axum::Server::from_tcp(listener)
+                    .expect("Unable to start snapshot server")
+                    .serve(app.into_make_service())
+                    .await
+                    .unwrap();
+            });
+        } else {
+            axum::Server::from_tcp(listener)
+                .expect("Unable to start snapshot server")
+                .serve(app.into_make_service())
+                .await
+                .unwrap();
+        }
+        Self {
+            socket_address: local_address,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct Snapshot {
+    request: Request,
+    response: Response,
+}
+
+impl TryFrom<Snapshot> for http::Response<RouterBody> {
+    type Error = ();
+
+    fn try_from(snapshot: Snapshot) -> Result<Self, Self::Error> {
+        let mut response = http::Response::builder().status(snapshot.response.status);
+        if let Some(headers) = response.headers_mut() {
+            for (name, values) in snapshot.response.headers.into_iter() {
+                if let Ok(name) = HeaderName::from_str(&name.clone()) {
```

Comment:
Is it OK to throw away unsupported header names? I would assume probably, but maybe have it warn / error / something if configured incorrectly since it would probably cause confusion when testing if things are just missing from the snapshots.